staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -679,7 +703,7 @@ func (s *Controller) nodeSyncLoop() {
 	if nodeSlicesEqualForLB(newHosts, s.knownHosts) {
 		// The set of nodes in the cluster hasn't changed, but we can retry
 		// updating any services that we failed to update last time around.
-		s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, newHosts)
+		s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, workers)
```

Comment:
This check doesn't work anymore.

Given that you list nodes again in `nodeSyncService` it may happen that:

- you have nodes A and B at this point
- in the meantime node C is added
- you call nodeSyncService with nodes A, B and C (so you're programmed for all 3 nodes)
- in the meantime node C is deleted
-  when you call it here the next time, you will not reprogram all services, because we will think that nothing changed

Reply:
I think the current logic would handle your case. Here is how: 

1. node A and B at one point. 
2. node C got added. Triggers `nodeSyncInternal`
3. nodeSyncInternal would record knownHosts= node A B C at one point
4. node C got deleted, Triggers nodeSyncLoop but got cached in the size 1 buffered channel
5. Meanwhile as step 4. happens, nodeSyncService updates LB with node A, B, C
6. in the next nodeSyncInternal triggered on step 4. knownHosts (node A, B, C) is not equal to newHosts (node A, B). Hence, LB will get a full resync on backend nodes.

Reply:
I find this logic pretty vulnerable as well. I have not find a good way to modify it so I just keep it as is. 

There is problematic scenario similar to what you pointed out:
Node A, B, C at one point. Node C got preempted and come back up very quickly. So cloud LB control plane underneath will deprogram C without service controller calling the API explicitly. But when nodeSync runs, all node A B C are present and nothing changed, hence node sync got skipped. If this happens enough times, all node A, B, C will be eventually moved out of the LB backends and cause outages.

Reply:
The current logic is optimized for avoiding the LB API calls as much as possible.

Reply:
ideally we want only one logic - to update all LBs each time there is a nodes change and get rid of periodic sync, correct?
 The problem here is that we cannot set s.knownHosts to an accurate value since each LB update used a different list most likely.

Should we change this function to perform node updates for all services all the time and increase the interval between periodic updates?

Reply:
Ok - so the reasoning is that it's eventually consistent, because we always program LBs with the state that is at least as fresh as `knowHosts`. That's a good argument, but maybe please document it explicitly.

Reply:
Done. Added detail comment on the end to end implications of recording the known states above.

Reply:
Actually no - I was right that it doesn't work now.
Let me give a more precise example (now that I recall what I had on my mind):

t0: node A is added, item inserted to nodeSyncCh
t1: nodeSyncInternal calls, newHosts={A}
t2. updateLoadBalancerHost is called
t3: node B is added, item inserted to nodeSyncCh
t4; nodes listed from cache - got {A,B}
t5: programming LB with {A, B}
t6: node B is deleted, trying to put to nodeSyncCh, already full
t7: programming end
t8: nodeSyncInternal called again
t9: Hosts ={A} [ which is the same as knownHosts => skip

But LB is programmed for {A,B}. And we won't recover until future change in node set.

Reply:
Good catch. let us discuss how to address this. I have 2 potential approaches. 

1. Beef up `triggerNodeSync` and let it compare the knownHosts vs. newHosts. Hence `triggerNodeSync` basically decides if nodeSync runs a full sync across all services or just do retry. 
2. Use a queue to contain services that needs nodeSync. When ever there is `triggerNodeSync` happens and node changes, enqueue all services into the queue. Plus the queue would need to handle retries and etc..

the 1st one is a simpler change, but we need to sort out if it contains any other corner cases.

2nd one is a bigger change to the nodesync logic today. Pro: it will be more consistent with the current controller pattern. Cons: Hard to collect latency metrics for a full sync on all services

Reply:
As discussed offline - (2) is much larger and much more risky change (maybe even deserving a KEP), so I'm fine with proceeding with (1).

Reply:
Done with (1). 

Ready for another round